functions to store and load metafiles for the database

Cranc opened the pull request:
Update to mod.rs 
includes:
- table functions
- column functions
- Custom Errors
- Store and load
- added enums DataTypes and FileModes

Commit:
Add table_test.rs for testing create and write function of storage table
Add save function to Table to save information in binary to Metafiles
Add open_file function to Table currently opens a file with access to read and write
Add create_new to Table to create test files
Add FileMode Enum - no function yet
Add new datatypes to Table (engine_id, version_nmbr, magic_nmbr, column_nmbr)

Commit:
Add some documentation and comments
Add Column functionality
Add self-save function to Column and enabled it in save function of table
Add simple Datatype for Debug reasons
Add configuration for open_file

Commit:
Merge branch 'dev' of https://github.com/OsnaCS/uosql-server into dev

Commit:
Add function to table::load
Change the workings of encoding and decoding
some cleaning
Column::save_to_file is now deprecated

Commit:
Add Database class to create Databases
Add Path usage to Table::open_file

Commit:
Add Database Error Enum

Change all Error Results to custom ones

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +31,128 @@ pub trait Engine {
+
+impl Database {
+    pub fn new_database(database_name: &str) -> Database{
+        Database{name: database_name.to_string()}
```
Just a note: `into()` instead of `to_string()` would work here.

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +31,128 @@ pub trait Engine {
+pub struct Table {
+    engine_id: u8,
+    version_nmbr: u8,
+    magic_nmbr: u8,
```
It doesn't really make sense the save the magic number in each table object. The magic number will never change and is just for file verification. So you could use a constant (`const MAGIC_NUMBER: u8 = 180`). Also: Usually magic numbers are 4 or 8 bytes wide.

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +31,128 @@ pub trait Engine {
-        Table
+        let mut file = try!(Self::open_file(database,table,FileMode::LoadDefault));
+
+        let data: Table = decode_from(&mut file, SizeLimit::Infinite).unwrap();
```
Spotted an `unwrap`. `try!` should work here...

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +31,128 @@ pub trait Engine {
+        Ok(())
+    }
+
+    pub fn create_new() -> Table {
```
I assume this function is currently just for testing purposes... But some when you will provide a function to create a new `Table` object. This function should either be `fn new(name: &str, engine_id: u8)` or -- if you want to provide a blanco object -- you should implement the `Default` trait from `std`.

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +31,128 @@ pub trait Engine {
+
+    fn open_file(database: &str, table: &str, mode: FileMode) -> Result<File, DatabaseError> {
+        //create new file or open new one
+        let st = String::from(format!("{}/{}",database,table));
```
`format!` already returns a `String`, so you should be able to remove the `String::from()`.

LukasKalbertodt commented:
- Braces `{` and `}` should never touch anything else. They should always be separated by a whitespace from everything else.
- A comma always touches the thing on the left, but never the following thing. So `{ engine_id: 3 , version_nmbr:  1,` is a space too much and `encode_into(&self,&mut file,SizeLimit::Infinite)` are two spaces too little.
- When using comments, please put a whitespace between the first char and the second `/` (e.g. not `//hello`, but `// hello`). You can fix all your comments with this regex search (`//([^/])`) and replace (`// \1`).

I also made some line comments. After fixing those issues, I will merge it ;)

Commit:
Add ByteOrder Errors to replace Io Error and some cleaning

Change Magic Number to const
Add write and controll for Magic number on load and save
Add final create_new function for Table and Column
Add skeleton for add_columns and remove_columns in table

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +38,172 @@ pub trait Engine {
+#[repr(u8)]
+#[derive(Clone,Copy,Debug,RustcDecodable, RustcEncodable)]
+pub enum DataType{ Integer = 1, Float = 2,}
+impl DataType{
```
Whitespace before `{`

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +38,172 @@ pub trait Engine {
+
+impl Database {
+    pub fn new_database(database_name: &str) -> Database {
+        Database{name: database_name.to_string()}
```
Whitespaces around { }

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +38,172 @@ pub trait Engine {
+
+impl Default for Table {
+    fn default() -> Table
+    {
```
{ in line above: `fn default() -> Table {`

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +38,172 @@ pub trait Engine {
+    fn default() -> Table
+    {
+        let mut t: Vec<Column> = Vec::new();
+        t.push(Default::default());
```
You could write

``` rust
let mut t = Vec::new();
t.push(Column::default());
```

instead... but not necessary.

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +38,172 @@ pub trait Engine {
+
+    pub fn create_new(engine: u8, name: &str) -> Table {
+
+        Table { name: name.to_string(), engine_id: engine, version_nmbr: 1, column_nmbr: 0, columns: Vec::new() }
```
This line is much too long, please split it into multiple lines. Like:

``` rust
Table { 
    name: name.to_string(), 
    engine_id: engine, 
    version_nmbr: 1, 
    column_nmbr: 0, 
    columns: Vec::new()
}
```

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +38,172 @@ pub trait Engine {
+        // call for open file
+        let mut file = try!(Self::open_file(database,table,FileMode::SaveDefault));
+        try!(file.write_u64::<BigEndian>(MAGIC_NUMBER));//MAGIC_NUMBER
+        try!(encode_into(&self,&mut file,SizeLimit::Infinite));
```
Here and in the line two above, there are whitespaces after commas missing.

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +38,172 @@ pub trait Engine {
+
+impl Default for Column {
+    fn default() -> Column
+    {
```
{ in line above: `fn default() -> Column {`

LukasKalbertodt commented on src/storage/mod.rs:
```diff
@@ -19,22 +38,172 @@ pub trait Engine {
+
+impl Column {
+    pub fn create_new(name: &str, dtype: DataType) -> Column {
+        Column{ name: name.to_string(), data_type: dtype.clone() }
```
Whitespace before {

Commit:
Change Magic Number and some code formatting

Commit:
More Code Formatting
